<?xml version='1.0' encoding='UTF-8'?>
<root version="2.3.8">
 <tabbed_widget name="Main Window" parent="main_window">
  <plotmatrix columns="1" rows="2" tab_name="Robot XY Position">
   <plot col="0" row="0">
    <range bottom="-0.002390" top="0.098010" left="16.600000" right="36.770000"/>
    <limitY/>
    <curve name="/goal_position_mb/x" R="26" G="201" custom_transform="noTransform" B="56"/>
    <curve name="/got_pose/pose/pose/position/x" R="31" G="119" custom_transform="noTransform" B="180"/>
    <curve name="/ground_truth/pose/position/x" R="214" G="39" custom_transform="noTransform" B="40"/>
    <curve name="/odom/pose/position/x" R="23" G="190" custom_transform="noTransform" B="207"/>
    <curve name="/odometry/filtered_odom/pose/position/x" R="188" G="189" custom_transform="noTransform" B="34"/>
    <curve name="/poseupdate/pose/pose/position/x" R="241" G="76" custom_transform="noTransform" B="193"/>
    <curve name="/slam_out_pose/pose/position/x" R="148" G="103" custom_transform="noTransform" B="189"/>
    <transform value="noTransform"/>
   </plot>
   <plot col="0" row="1">
    <range bottom="-0.010867" top="0.010378" left="16.600000" right="36.770000"/>
    <limitY/>
    <curve name="/goal_position_mb/y" R="26" G="201" custom_transform="noTransform" B="56"/>
    <curve name="/got_pose/pose/pose/position/y" R="31" G="119" custom_transform="noTransform" B="180"/>
    <curve name="/ground_truth/pose/position/y" R="214" G="39" custom_transform="noTransform" B="40"/>
    <curve name="/odom/pose/position/y" R="23" G="190" custom_transform="noTransform" B="207"/>
    <curve name="/odometry/filtered_odom/pose/position/y" R="188" G="189" custom_transform="noTransform" B="34"/>
    <curve name="/poseupdate/pose/pose/position/y" R="26" G="201" custom_transform="noTransform" B="56"/>
    <curve name="/slam_out_pose/pose/position/y" R="148" G="103" custom_transform="noTransform" B="189"/>
    <transform value="noTransform"/>
   </plot>
  </plotmatrix>
  <currentPlotMatrix index="0"/>
 </tabbed_widget>
 <use_relative_time_offset enabled="1"/>
 <!-- - - - - - - - - - - - - - - -->
 <!-- - - - - - - - - - - - - - - -->
 <Plugins>
  <plugin ID="DataLoad CSV">
   <default time_axis=""/>
  </plugin>
  <plugin ID="DataLoad ROS bags">
   <use_header_stamp value="false"/>
   <use_renaming_rules value="true"/>
   <discard_large_arrays value="true"/>
   <max_array_size value="100"/>
  </plugin>
  <plugin ID="DataLoad ULog"/>
  <plugin ID="ROS Topic Subscriber">
   <use_header_stamp value="false"/>
   <use_renaming_rules value="true"/>
   <discard_large_arrays value="true"/>
   <max_array_size value="100"/>
  </plugin>
  <plugin ID="ROS /rosout Visualization" status="idle"/>
  <plugin ID="ROS Topic Re-Publisher" status="idle"/>
 </Plugins>
 <!-- - - - - - - - - - - - - - - -->
 <previouslyLoaded_Datafiles/>
 <previouslyLoaded_Streamer name="ROS Topic Subscriber"/>
 <!-- - - - - - - - - - - - - - - -->
 <customMathEquations/>
 <snippets>
  <snippet name="1st_derivative" language="JS">
   <global>var prevX = 0
var prevY = 0</global>
   <equation>dx = time - prevX
dy = value - prevY
prevX = time
prevY = value

return dy/dx</equation>
  </snippet>
  <snippet name="1st_order_lowpass" language="JS">
   <global>var prevY = 0
var alpha = 0.1</global>
   <equation>prevY = alpha * value + (1.-alpha) * prevY

return prevY</equation>
  </snippet>
  <snippet name="2D_velocity" language="JS">
   <global>var prev_x = 0
var prev_y = 0
var prev_t = 0</global>
   <equation>X = $$your_odometry/position/x$$
Y = $$your_odometry/position/y$$

var dist = sqrt( (X-prev_x)*(X-prev_x) + (Y-prev_y)*(Y-prev_y) )
var dT = time - prev_t

prev_x = X
prev_y = Y
prev_t = time

return dist / dT</equation>
  </snippet>
  <snippet name="average_two_curves" language="JS">
   <global></global>
   <equation>a = $$PLOT_A$$
b = $$PLOT_B$$

return (a+b)/2</equation>
  </snippet>
  <snippet name="integral" language="JS">
   <global>var integral = 0</global>
   <equation>integral += value
return integral</equation>
  </snippet>
  <snippet name="rad_to_deg" language="JS">
   <global></global>
   <equation>return value*180/3.1417</equation>
  </snippet>
  <snippet name="remove_offset" language="JS">
   <global>var is_first = true
var first_value = 0</global>
   <equation>if (is_first)
{
  is_first = false
  first_value = value
}

return value - first_value</equation>
  </snippet>
  <snippet name="yaw_from_quaternion" language="JS">
   <global>// source: https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles

function quaternionToYaw(x, y, z, w)
{
  // yaw (z-axis rotation)
  t1 = 2.0 * (w * z + x * y);
  t2 = 1.0 - 2.0 * (y * y + z * z);
  yaw = Math.atan2(t1, t2);

  return yaw
}</global>
   <equation>return quaternionToYaw(x, y, z, w);</equation>
  </snippet>
 </snippets>
 <!-- - - - - - - - - - - - - - - -->
</root>

